src/osd/OSD.cc

```diff
@@ -8308,63 +8311,72 @@ void OSD::handle_osd_map(MOSDMap *m)
-  for (auto& [added_map_epoch, added_map] : added_maps) {
-    if (!lastmap) {
-      if (!(lastmap = service.try_get_map(added_map_epoch - 1))) {
-        dout(10) << __func__ << " can't get previous map " << added_map_epoch - 1
-                 << " probably first start of this osd" << dendl;
-        continue;
-      }
+  OSDMapRef current_lastmap = added_maps[last];
+  OSDMapRef previous_lastmap;
+
+  if (superblock.maps.empty()) {
+    dout(10) << __func__ << " no maps stored, this is probably "
+             << "the first start of this osd" << dendl;
+    previous_lastmap = added_maps[first];
+  } else {
+    if (first > superblock.get_newest_map() + 1) {
+      ceph_assert(first == superblock.cluster_osdmap_trim_lower_bound);
+      dout(10) << __func__ << " can't get previous map " << first - 1
+               << " first start of this osd after a map gap" << dendl;
+    }
+    if (!(previous_lastmap = service.try_get_map(superblock.get_newest_map()))) {
+      // this is unexpected, either fallback to the first added map:
+      // previous_lastmap = added_maps[first];
+      // or abort:
+      ceph_abort();
     }
-    ceph_assert(lastmap->get_epoch() + 1 == added_map->get_epoch());
-    for (auto& [pool_id, pg_pool] : lastmap->get_pools()) {
-      if (!added_map->have_pg_pool(pool_id)) {
-	pg_num_history.log_pool_delete(added_map_epoch, pool_id);
-	dout(10) << __func__ << " recording final pg_pool_t for pool "
-		 << pool_id << dendl;
-	// this information is needed by _make_pg() if have to restart before
-	// the pool is deleted and need to instantiate a new (zombie) PG[Pool].
-	ghobject_t obj = make_final_pool_info_oid(pool_id);
-	bufferlist bl;
-	encode(pg_pool, bl, CEPH_FEATURES_ALL);
-	string name = lastmap->get_pool_name(pool_id);
-	encode(name, bl);
-	map<string,string> profile;
-	if (lastmap->get_pg_pool(pool_id)->is_erasure()) {
-	  profile = lastmap->get_erasure_code_profile(
-	    lastmap->get_pg_pool(pool_id)->erasure_code_profile);
-	}
-	encode(profile, bl);
-	t.write(coll_t::meta(), obj, 0, bl.length(), bl);
-      } else if (unsigned new_pg_num = added_map->get_pg_num(pool_id);
-		 new_pg_num != pg_pool.get_pg_num()) {
-	dout(10) << __func__ << " recording pool " << pool_id << " pg_num "
-		 << pg_pool.get_pg_num() << " -> " << new_pg_num << dendl;
-	pg_num_history.log_pg_num_change(added_map_epoch, pool_id, new_pg_num);
+  }
+
+  // 1) Check if a pool was deleted
+  for (auto& [pool_id, pg_pool] : previous_lastmap->get_pools()) {
+    if (!current_lastmap->have_pg_pool(pool_id)) {
+      pg_num_history.log_pool_delete(last, pool_id);
+      dout(10) << __func__ << " recording final pg_pool_t for pool "
+               << pool_id << dendl;
+      // this information is needed by _make_pg() if have to restart before
+      // the pool is deleted and need to instantiate a new (zombie) PG[Pool].
+      ghobject_t obj = make_final_pool_info_oid(pool_id);
+      bufferlist bl;
+      encode(pg_pool, bl, CEPH_FEATURES_ALL);
```

Comment:
If previous_lastmap->get_epoch() is 10 and current_lastmap->get_epoch() is 20, the pool might actually have been removed in epoch 19.  If that's the case, it's possible that the pg_pool structure in epoch 18 is different from the one in 10.  Does it not matter that we have the most recent pg_pool prior to removal?

Reply:
Added last  2 commits which _should_ explain why it doesn't matter. If you find that still to be risky, I can omit the change from this PR and proceed with the mapgap issue fix and refactors only.

Reply:
See my other comment.  I see users other than the loop bounds and I'm not convinced that it's harmless.  Seems like it wouldn't be all that complex to fill in pg_num_history accurately as this function already gets the maps it would need.